sdk/include/ds/linked_list.h

```diff
@@ -0,0 +1,642 @@
+	 * returns `true`, the search stops early and returns `true`; otherwise,
+	 * search returns `false`.  To (side-effectfully) visit every node in the
+	 * span, have `f` always return false.
+	 */
+	template<cell::HasPrimOps Cell, typename F>
+	__always_inline bool search(Cell *from, Cell *to, F f)
+	{
+		Cell *elem;
+		for (elem = from; elem != to; elem = elem->cell_next())
+		{
+			if (f(elem))
+			{
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Search through all elements of a ring *except* `elem`.  If `elem` is the
+	 * sentinel of a ring, then this is, as one expects, a `search` over all
+	 * non-sentinel memebers of the ring.
+	 */
+	template<cell::HasPrimOps Cell, typename F>
+	__always_inline bool search(Cell *elem, F f)
+	{
+		return search(static_cast<Cell *>(elem->cell_next()), elem, f);
+	}
+
+	/**
+	 * Convenience wrapper for a sentinel cons cell, encapsulating some common
+	 * patterns.
+	 */
+	template<cell::HasPrimOpsInit Cell_>
+	struct Sentinel
+	{
+		using Cell = Cell_;
+
+		/**
+		 * The sentinel node itself.  Viewing the ring as a list, this
+		 * effectively serves as pointers to the head (next) and to the tail
+		 * (prev) of the list.  Unlike more traditional nullptr-terminated
+		 * lists, though, here, the sentinel participates in the ring.
+		 *
+		 * This is marked `cheri_no_subobject_bounds` because some of our cons
+		 * cell implementations use pointer proxies that rely on the bounds
+		 * provided by `this` (which, in turn, is likely to be
+		 * `cheri_no_subobject_bounds`)
+		 */
+		Cell sentinel __attribute__((__cheri_no_subobject_bounds__));
+
+		__always_inline void init()
+		{
+			sentinel.cell_init();
+		}
+
+		__always_inline bool is_empty()
+		{
+			return linked_list::is_singleton(&sentinel);
+		}
+
+		__always_inline void append(Cell *elem)
+		{
+			linked_list::insert_before(&sentinel, elem);
+		}
+
+		__always_inline void append_new(Cell *elem)
+		{
+			linked_list::insert_new_before(&sentinel, elem);
+		}
+
+		__always_inline void prepend(Cell *elem)
+		{
+			linked_list::insert_before(elem, &sentinel);
+		}
+
+		__always_inline Cell *first()
+		{
+			return sentinel.cell_next();
+		}
+
+		__always_inline Cell *last()
+		{
+			return sentinel.cell_prev();
+		}
+
+		__always_inline bool last_is_eq(Cell *p)
+		{
+			return p->cell_next_is_eq(&sentinel);
+		}
+
+		__always_inline Cell *unsafe_take_first()
+		{
+			Cell *f = sentinel.cell_next();
+			linked_list::unsafe_remove_link(&sentinel, f);
+			return f;
+		}
+
+		__always_inline Cell *take_all()
+		{
+			auto p = linked_list::unsafe_remove(&sentinel);
+			sentinel.cell_init();
+			return p;
+		}
+
+		template<typename F>
+		__always_inline bool search(F f)
+		{
+			return linked_list::search(&sentinel, f);
+		}
+	};
+
+	namespace cell
+	{
+
+		/** Cons cell using two pointers */
+		class Pointer
+		{
+			Pointer *prev, *next;
+
+			public:
+			__always_inline void cell_init()
```

Comment:
Why is this not a constructor?

Reply:
Because we might want to reset an existing object and, in principle, that object might be bit-packing other state into its representation, so we can't just placement-new a new cell over the old one.  I've added a comment.

Reply:
It makes me uncomfortable that we can have objects that are not in a valid state (especially coming from a Haskell programmer!), rather than requiring that they are constructed in a valid state.  If we can use placement new and ensure that we cannot construct a pointer to a list element without having first placed it in a well-defined state, that would be somewhat less fragile.

If you are concerned about the bit packing, perhaps you could add factory methods that the caller can use that wrap placement new (e.g. `construct_in_place(void*ptr)`)?